### ISSUE TITLE
Proposal: reduce the default allowed tags

### ISSUE THREAD

**Issue by vanillajonathan** [mganss/HtmlSanitizer#350]
Remove all legacy/deprecated tags such as `acronym`, `big`, `center`, `dir`, `keygen`, `menuitem` and `strike`.

Remove non-semantic presentational HTML4 tags such as `b`, `i`, `u` and `tt`.

Remove tags dealing with forms such as `button`, `datalist`, `fieldset`, `form`, `input`, `keygen`, `label`, `legend`, `optgroup`, `option`, `output` and `select`. Remove UI tags such as `meter`, `progress`.

Remove the `main` tag just as the `html` and `body` tags aren't allowed.

**Comment by mganss** [comment]
I'd rather not mess with the default set for backward compatibility reasons. But perhaps it would be a good idea to provide static readymade sets for different use cases like https://github.com/rgrove/sanitize#configuration (restricted, basic, relaxed). What do you think?

**Comment by mganss** [comment]
I agree in that we could change the defaults and up the major version. I'm unsure though what exactly we should use as guidance when selecting the default set. The original motivation was to eliminate only those elements that carry XSS potential. For example, https://github.com/rgrove/sanitize allows `b` and `i` even in the most restricted configuration.

**Comment by vanillajonathan** [comment]
Right now it is possible to do some undesirable things such as using `display: none` or `visibility: hidden` to hide spam, set the `min-width`, `width`, `min-height` and `height` to excessive values like 9000px, etc.

I think HtmlSanatizer is commonly used for comments, forum posts, blog posts and to author content in a CMS.

I am not even sure there ought to be a default set at all. If a default set should exist, I would think it should be heavily restricted to perhaps just semantic elements for user-submitted comment; this would include `strong` and `em` but not `h1` .. `h6`.

**Comment by tiesont** [comment]
I think any decision about changing the default needs to be based on deciding what is **in** scope for HtmlSanitizer and what **isn't**. Arbitrarily deciding what constitutes valid HTML feels out of scope, to me; as @mganss says, the main motivator for HtmlSanitizer is, well, sanitizing HTML.

I do like the idea of having static predefined sets that can be passed to the various options.

**Comment by vanillajonathan** [comment]
Good point @tiesont!

Maybe move the default to a static class called `Presets` and change the example in the documentation to:
```cs
var sanitizer = new HtmlSanitizer();
sanitizer.AllowedTags = Presets.AllXssSafeTags;
sanitizer.AllowedAttributes = Presets.AllXssSafeAttributes;
sanitizer.AllowedCssProperties = Presets.AllXssSafeCssProperties;
sanitizer.AllowedAtRules = Presets.AllXssSafeAtRules;
sanitizer.AllowedSchemes = Presets.AllXssSafeSchemes;
sanitizer.UriAttributes = Presets.AllXssSafeUriAttributes;
var html = "<p>Hello world!</p>";
var sanitized = sanitizer.Sanitize(html, "https://www.example.com");
```

Though this would kind of break the expectation that a class when newed up should be fully instantiated and ready to be used. But maybe it is fully initialized since the `Sanitize` method _can_ be called, although it will sanitize **everything** since nothing would be allowed. To prevent this, the constructor could be changed to require the parameters but maybe that would be awkward too.

**Comment by tiesont** [comment]
You'd probably want to follow the pattern used by a lot of .NET Core utilities - rather than pass in multiple parameters, pass in a configuration/options object (maybe even allow a factory method). The default constructor would just use the existing defaults.

**Comment by vanillajonathan** [comment]
So something like:
```cs
var options = new HtmlSanitizerOptions
{
    AllowedTags = Presets.AllXssSafeTags;
    AllowedAttributes = Presets.AllXssSafeAttributes;
    AllowedCssProperties = Presets.AllXssSafeCssProperties;
    AllowedAtRules = Presets.AllXssSafeAtRules;
    AllowedSchemes = Presets.AllXssSafeSchemes;
    UriAttributes = Presets.AllXssSafeUriAttributes;
};

var sanitizer = new HtmlSanitizer(options);
```

**Comment by tiesont** [comment]
Exactly.

**Comment by tiesont** [comment]
Would it make any sense to also provide a constructor which accepts a `Func<HtmlSanitizerOptions>`?

You might use it like so:

```c#
var sanitizer = new HtmlSanitizer(() => {
    var options = new HtmlSanitizerOptions(Presets.Strict);

    // do some other stuff with options as necessary
    // --> assuming actions that can't be called with initializer syntax

    return options;
});
```

This particular example doesn't really have any benefit over just instantiating the options object and then passing it as a parameter, but I'd assume it would allow a factory delegate of some sort to be passed, too?

**Comment by vanillajonathan** [comment]
Sounds reasonable. It would be an `Action`, so it would be `Action<HtmlSanitizerOptions>`.

I've been hesitating on the presets. It would be singular not plural because plural are used for bitflags enum while singular for normal enums. `Preset` is a bit ambiguous, so I was thinking `HtmlSanitizerPreset` or `HtmlSanitizerOptionsPreset` but that is a bit long though.

Right now we only have the default preset so the enum would just have one variant. It could be called `Default` but if the ambition is to move towards a more reduced defaults then maybe it ought to be named `Relaxed` or `Loose` or something instead.

With pull request #370 the constructor taking arguments is removed and a new parameterless constructor is introduced which configures the `HtmlSanitizer` with the current defaults that is already used today.

If taking in an enum for presets as input parameter then the class would grow with a new if condition for every variant of the enum. Another approach would to use a factory to keep all that `if` logic within a factory class. The factory could have take in an enum, or instead of an enum it could just have methods instead. Example `HtmlSanitizerOptionsFactory.CreateRelaxed()`.

**Comment by tiesont** [comment]
@vanillajonathan I think I'm a little confused - what are you referring to, with respect to enums?

**Comment by mganss** [comment]
@tiesont Re the `Action` parameter, I'm not sure I understand the use case. I know this pattern from `IServiceCollection` where you configure a DI container with an `Action` which gets called at a later time when the container is built. But where's the need to postpone execution of the `Action` in the scope of `HtmlSanitizer`? Or is there another reason I'm missing?

@vanillajonathan I favor the second approach for the reasons you have mentioned. Why not have static members like so?

```c#
public class HtmlSanitizerOptions
{
    public static readonly HtmlSanitizerOptions Relaxed = CreateRelaxed();

    private static HtmlSanitizerOptions CreateRelaxed()
    {
        // ...
    }
}
```

**Comment by vanillajonathan** [comment]
Well the enum could be on the `HtmlSanitizer` class, but I was thinking of having on the `HtmlSanitizerOptions` class because what is what is done by [`JsonSerializerOptions`](https://docs.microsoft.com/en-us/dotnet/api/system.text.json.jsonserializeroptions?view=net-6.0).

But I was thinking of ditching the enum altogether and instead using a separate factory class with a factory method for each preset. This would avoid a long method with many if/else statements or switch.

@mganss Yeah static members on the `HtmlSanitizerOptions` is an alternative too, that would be factory methods on the options class instead of having a separate factory class. The advantage of having the factory method on the options class is that it is arguably more discoverable than a separate factory class, and arguably more cohesive and local. An argument for a separate factory class would be that it separates things. I like your idea though.

**Comment by mganss** [comment]
Perhaps the presets should be `ImmutableHashSet`s. It feels wrong that the caller could change the static defaults.
